

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=&#34;auto&#34;>



<head>
  <meta charset="UTF-8">
  <link rel="apple-touch-icon" sizes="76x76" href="/img/logo/genetics.png">
  <link rel="icon" href="/img/logo/genetics.png">
  <meta name="viewport"
        content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="description" content="Weber的博客">
  <meta name="author" content="Weber_Kang">
  <meta name="keywords" content="">
  
  <title>STAR和BWA比对软件使用；samtools使用 - Weber&#39;s Blog</title>

  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" />


  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/github-markdown-css@4.0.0/github-markdown.min.css" />
  <link  rel="stylesheet" href="/lib/hint/hint.min.css" />

  
    
    
      
      <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@10.4.0/styles/github-gist.min.css" />
    
  

  
    <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.5.7/dist/jquery.fancybox.min.css" />
  



<!-- 主题依赖的图标库，不要自行修改 -->

<link rel="stylesheet" href="//at.alicdn.com/t/font_1749284_ba1fz6golrf.css">



<link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_kmeydafke9r.css">


<link  rel="stylesheet" href="/css/main.css" />

<!-- 自定义样式保持在最底部 -->


  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    var CONFIG = {"hostname":"example.com","root":"/","version":"1.8.11","typing":{"enable":true,"typeSpeed":100,"cursorChar":"_","loop":false},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"right","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"copy_btn":true,"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"baidu":null,"google":null,"gtag":null,"tencent":{"sid":null,"cid":null},"woyaola":null,"cnzz":null,"leancloud":{"app_id":null,"app_key":null,"server_url":null}},"search_path":"/local-search.xml"};
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
<meta name="generator" content="Hexo 5.4.0"></head>


<body>
  <header style="height: 70vh;">
    <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand"
       href="/">&nbsp;<strong>Weber Blog</strong>&nbsp;</a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/">
                <i class="iconfont icon-home-fill"></i>
                首页
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/">
                <i class="iconfont icon-archive-fill"></i>
                归档
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/">
                <i class="iconfont icon-category-fill"></i>
                分类
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/">
                <i class="iconfont icon-tags-fill"></i>
                标签
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/">
                <i class="iconfont icon-user-fill"></i>
                关于
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" data-toggle="modal" data-target="#modalSearch">&nbsp;<i
                class="iconfont icon-search"></i>&nbsp;</a>
          </li>
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self">&nbsp;<i
                class="iconfont icon-dark" id="color-toggle-icon"></i>&nbsp;</a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

    <div class="banner" id="banner" parallax=true
         style="background: url('/img/banner/banner.jpg') no-repeat center center;
           background-size: cover;">
      <div class="full-bg-img">
        <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
          <div class="page-header text-center fade-in-up">
            <span class="h2" id="subtitle" title="STAR和BWA比对软件使用；samtools使用">
              
            </span>

            
              <div class="mt-3">
  
  
    <span class="post-meta">
      <i class="iconfont icon-date-fill" aria-hidden="true"></i>
      <time datetime="2021-07-12 09:36" pubdate>
        2021年7月12日 上午
      </time>
    </span>
  
</div>

<div class="mt-1">
  
    
    <span class="post-meta mr-2">
      <i class="iconfont icon-chart"></i>
      3.6k 字
    </span>
  

  
    
    <span class="post-meta mr-2">
      <i class="iconfont icon-clock-fill"></i>
      
      
      48
       分钟
    </span>
  

  
  
</div>

            
          </div>

          
        </div>
      </div>
    </div>
  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="d-none d-lg-block col-lg-2"></div>
    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div class="py-5" id="board">
          <article class="post-content mx-auto">
            <!-- SEO header -->
            <h1 style="display: none">STAR和BWA比对软件使用；samtools使用</h1>
            
            <div class="markdown-body">
              <h1 id="介绍"><a href="#介绍" class="headerlink" title="介绍"></a>介绍</h1><p><code>BWA</code>和<code>STAR</code>作为ENCODE的御用软件,分别用于DNA-seq和RNA-seq</p>
<blockquote>
<p>DNA-seq和RNA-seq区别，由于RNA形成中有外显子，区别显然在与区别仅在于是否会考虑跨外显子的比对（即：是否会将没有比对上的reads劈开，对劈开后的两部分再次比对）。</p>
<ul>
<li>DNA-seq：bowtie；bowtie2；BWA</li>
<li>RNA-seq：STAR；HISAT2；Tophat<br>这里我们主要介绍ENCODE的御用软件<br>  <code>samtools</code>是一个用于操作sam和bam文件的工具合集，和上面BWA一样都是大佬李恒开发的。<del><strong>题外话；开发高效软件只要用你的软件就算引用这引用真的高。</strong></del></li>
</ul>
</blockquote>
<h1 id="使用"><a href="#使用" class="headerlink" title="使用"></a>使用</h1><h2 id="STAR"><a href="#STAR" class="headerlink" title="STAR"></a>STAR</h2><h3 id="下载安装"><a href="#下载安装" class="headerlink" title="下载安装"></a>下载安装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">wget -c https://github.com/alexdobin/STAR/archive/2.7.3a.tar.gz<br>tar -xvzf 2.7.3a.tar.gz<br><span class="hljs-built_in">cd</span> STAR-2.7.3a<br></code></pre></td></tr></table></figure>

<h3 id="构建index"><a href="#构建index" class="headerlink" title="构建index"></a>构建index</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">STAR<br>--runThreadN NumberOfThreads<br>--runMode genomeGenerate<br>--genomeDir /path/to/genomeDir<br>--genomeFastaFiles /path/to/genome/fasta1 /path/to/genome/fasta2 ...<br>--sjdbGTFfile /path/to/annotations.gtf<br>--sjdbOverhang ReadLength-1<br></code></pre></td></tr></table></figure>

<hr>
<p><code>--runThreadN</code> ：设置线程数<br><code>--genomeDir</code> ：index输出的路径<br><code>--genomeFastaFiles</code> ：参考基因组序列<br><code>--sjdbGTFfile</code> ：参考基因组注释文件<br><code>--sjdbOverhang</code> ：这个是reads长度的最大值减1，默认是100<br><code>--sjdbFileChrStartEnd</code>：与注释(通过<code>--sjdbGTFfile</code>选项)一起生成新的基因组索引<br>第二遍映射。这需要对所有样本只做一次。</p>
<h3 id="进行mapping"><a href="#进行mapping" class="headerlink" title="进行mapping"></a>进行mapping</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash">STAR<br>--runThreadN NumberOfThreads<br>--genomeDir /path/to/genomeDir<br>--outSAMtype BAM SortedByCoordinate<br>--readFilesIn /path/to/read1 [/path/to/read2 ]<br>--readFilesCommand UncompressionCommand<br></code></pre></td></tr></table></figure>

<p><code>--twopassMode Basic</code>：使用two-pass模式进行reads比对。简单来说就是先按索引进行第一次比对，而后把第一次比对发现的新剪切位点信息加入到索引中进行第二次比对。<br><code>--quantMode TranscriptomeSAM GeneCounts</code>：将reads比对至转录本序列。<br><code>--runThreadN</code>：线程数。<br><code>--genomeDir</code>：索引目录。<br><code>--alignIntronMin</code>：最短的内含子长度。（根据GTF文件计算）<br><code>--alignIntronMax</code>：最长的内含子长度。（根据GTF文件计算）<br><code>--outSAMtype BAM SortedByCoordinate</code>：输出BAM文件并进行排序。<br><code>--sjdbOverhang</code>：reads长度减1。<br><code>--outSAMattrRGline</code>：ID代表样本ID，SM代表样本名称，PL为测序平台。在使用GATK进行SNP Calling时同一SM的样本可以合并在一起。<br><code>--outFilterMismatchNmax</code>：比对时允许的最大错配数。<br><code>--outSJfilterReads Unique</code>：对于跨越剪切位点的reads（junction reads)，只考虑跨越唯一剪切位点的reads。</p>
<p><code>--outSAMtype</code> ：表示输出默认排序的bam文件，类似于samtools sort（还有–outSAMtype BAM Unsorted和–outSAMtype BAM Unsorted SortedByCoordinate）</p>
<p><code>--outSAMmultNmax</code>：每条reads输出比对结果的数量。<br><code>--outFileNamePrefix</code>：输出文件前缀。<br><code>--outSAMmapqUnique 60</code>：将uniquely mapping reads的MAPQ值调整为60，满足下游使用GATK进行分析的需要。<br><code>--readFilesCommand</code>：对FASTQ文件进行操作。<br><code>--readFilesIn</code>：输入FASTQ文件的路径。</p>
<h3 id="两次mapping"><a href="#两次mapping" class="headerlink" title="两次mapping"></a>两次mapping</h3><blockquote>
<h4 id="STAR-2-pass-mode"><a href="#STAR-2-pass-mode" class="headerlink" title="STAR 2-pass mode"></a>STAR 2-pass mode</h4><p>为了发现更加灵敏的new junction，STAR建议使用2-pass mode，其能增加检测到的new junction数目，使得更多的splices reads能mapping到new junction。因此STAR先用一般参数做一遍mapping，收集检测到的junction信息，然后利用这已经annotated junction来做第二次mapping。</p>
</blockquote>
<p>首先做一遍常规的比对，结果中会生成一个SJ.out.tab文件，如上面所提到的，然后用<code>--sjdbFileChrStartEnd</code>参数将所有样品的SJ.out.tab文件作为输入的annotated junction进行第二次建index。</p>
<hr>
<p>上述方法original方法适用于多样本和单个样本的处理，但是如果是per-sample(单个样本？)的2-pass mapping，可以直接用<code>--twopassMode Basic</code>参数将第两步mapping中的make index省去，直接再mapping。</p>
<h3 id="如果不清楚请使用STAR帮助文档"><a href="#如果不清楚请使用STAR帮助文档" class="headerlink" title="如果不清楚请使用STAR帮助文档"></a>如果不清楚请使用STAR帮助文档</h3><p><a target="_blank" rel="noopener" href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf"><em>STARmanual</em></a></p>
<h2 id="BWA-Burrows-Wheeler-Alignment-Tool"><a href="#BWA-Burrows-Wheeler-Alignment-Tool" class="headerlink" title="BWA(Burrows-Wheeler-Alignment Tool)"></a>BWA(Burrows-Wheeler-Alignment Tool)</h2><blockquote>
<p>BWA 的三种模式</p>
<ul>
<li><p><strong>BWA-backtrack:</strong> 是用来比对 Illumina 的序列的，reads 长度最长能到 100bp。</p>
</li>
<li><p><strong>BWA-SW:</strong> 用于比对 long-read ，支持的长度为 70bp-1Mbp；同时支持剪接性比对。</p>
</li>
<li><p><strong>BWA-MEM:</strong> 推荐使用的算法，支持较长的read长度，同时支持剪接性比对（split alignments)，但是BWA-MEM是更新的算1法，也更快，更准确，且 BWA-MEM 对于 70bp-100bp 的 Illumina 数据来说，效果也更好些。</p>
</li>
</ul>
</blockquote>
<h3 id="下载安装-1"><a href="#下载安装-1" class="headerlink" title="下载安装"></a>下载安装</h3><ol>
<li><p>下载tar</p>
<p>下载BWA <a target="_blank" rel="noopener" href="https://github.com/lh3/bwa/releases">http://bio-bwa.sourceforge.net/bwa.shtml</a></p>
</li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment">#解压</span><br>$ tar -jxvf bwa-*.tar.bz2<br>$ <span class="hljs-built_in">cd</span> bwa-*；<br><span class="hljs-comment"># 编译BWA</span><br>$ make<br>$ <span class="hljs-built_in">echo</span> <span class="hljs-string">&#x27;PATH=$PATH:/path/bwa--*&#x27;</span> &gt;&gt; ~/.bashrc<br>$ <span class="hljs-built_in">source</span> ~/.bashrc<br></code></pre></td></tr></table></figure>

<ol start="2">
<li>git</li>
</ol>
 <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs bash">git <span class="hljs-built_in">clone</span> https://github.com/lh3/bwa.git<br><span class="hljs-built_in">cd</span> bwa<br>make<br><span class="hljs-comment">##command</span><br>Command: index         index sequences <span class="hljs-keyword">in</span> the FASTA format<br>         mem           BWA-MEM algorithm<br>         fastmap       identify super-maximal exact matches<br>         pemerge       merge overlapping paired ends (EXPERIMENTAL)<br>         aln           gapped/ungapped alignment<br>         samse         generate alignment (single ended)<br>         sampe         generate alignment (paired ended)<br>         bwasw         BWA-SW <span class="hljs-keyword">for</span> long queries<br>         shm           manage indices <span class="hljs-keyword">in</span> shared memory<br>         fa2pac        convert FASTA to PAC format<br>         pac2bwt       generate BWT from PAC<br>         pac2bwtgen    alternative algorithm <span class="hljs-keyword">for</span> generating BWT<br>         bwtupdate     update .bwt to the new format<br>         bwt2sa        generate SA from BWT and Occ<br></code></pre></td></tr></table></figure>

<h3 id="构建index-1"><a href="#构建index-1" class="headerlink" title="构建index"></a>构建index</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">bwa index [ –p prefix ] [ –a algoType ] &lt;in.db.fasta&gt;<br><span class="hljs-comment"># 根据reference genome data(e.g. ref.fa) 建立 Index File：</span><br>bwa index ref.fa -p genome        <span class="hljs-comment"># 可以不加-p genome，这样建立索引都是以ref.fa为前缀</span><br></code></pre></td></tr></table></figure>

<h3 id="进行mapping-1"><a href="#进行mapping-1" class="headerlink" title="进行mapping"></a>进行mapping</h3><h4 id="BWA-MEM-算法"><a href="#BWA-MEM-算法" class="headerlink" title="BWA-MEM 算法"></a>BWA-MEM 算法</h4><blockquote>
<p>该算法先使用 MEM(maximal exact matches) 进行 seeding alignments，再使用 SW(affine-gap Smith-Waterman) 算法进行 seeds 的延伸。BWA–MEM 算法执行局部比对和剪接性。可能会出现 query 序列的多个不同的部位出现各自的最优匹配，导致 reads 有多个最佳匹配位点。有些软件如 Picard’s markDuplicates 跟 mem 的这种剪接性比对不兼容,在这种情况下，可以使用 –M 选项来将 shorter split hits 标记为次优。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">bwa mem [options] ref.fa reads.fq [mates.fq]<br></code></pre></td></tr></table></figure>

<p>​     <code>-t</code>   INT 线程数，默认是1，增加线程数，会减少运行时间。<br> 　<code>-M</code>   <strong>将 shorter split hits 标记为次优，以兼容 Picard’s markDuplicates 软件。</strong><br> 　<code>-p</code>   若无此参数：输入文件只有1个，则进行单端比对；若输入文件有2个，则作为paired reads进行比对。若加入此参数：则仅以第1个文件作为输入(会忽略第二个输入序列文件，把第一个文件当做单端测序的数据进行比对)，该文件必须是read1.fq和read2.fa进行reads交叉的数据。<br> 　<code>-R</code>   STR 完整的read group的头部，可以用 ‘\t’ 作为分隔符， 在输出的SAM文件中被解释为制表符TAB. read group 的ID，会被添加到输出文件的每一个read的头部。<br> 　<code>-T</code>   INT   当比对的分值比 INT 小时，不输出该比对结果，这个参数只影响输出的结果，不影响比对的过程。<br> 　<code>-a</code>   将所有的比对结果都输出，包括 single-end 和 unpaired paired-end的 reads，但是这些比对的结果会被标记为次优。<br> 　<code>-Y</code>   对数据进行soft clipping, 当错配或者gap数过多比对不上时，会对序列进行切除，这里的切除并只是在比对时去掉这部分序列，最终输出结果中序列还是存在的，所以称为soft clipping。</p>
<h4 id="BWA其他算法"><a href="#BWA其他算法" class="headerlink" title="BWA其他算法"></a>BWA其他算法</h4><p>详情请见<a target="_blank" rel="noopener" href="http://bio-bwa.sourceforge.net/bwa.shtml"><em>BWAmanual</em></a></p>
<h2 id="SAMtools"><a href="#SAMtools" class="headerlink" title="SAMtools"></a>SAMtools</h2><blockquote>
<p>samtools是一个用于操作sam和bam文件的工具合集。能够实现二进制查看、格式转换、排序及合并等功能，结合sam格式中的flag、tag等信息，还可以完成比对结果的统计汇总。同时利用linux中的grep、awk等操作命令，还可以大大扩展samtools的使用范围与功能。包含有许多命令。</p>
</blockquote>
<h3 id="下载安装-2"><a href="#下载安装-2" class="headerlink" title="下载安装"></a>下载安装</h3><p><code>https://github.com/samtools/samtools/releases/</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash">mkdir ~/biosoft/samtools<br><span class="hljs-built_in">cd</span> ~/biosoft/samtools<br>wget -c https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2<br>tar jxvf samtools-1.9.tar.bz2<br>./configure --prefix=~/biosoft/samtools-1.9<br>make<br>make install<br><span class="hljs-built_in">echo</span> <span class="hljs-string">&#x27;export PATH=&quot;/home/vip47/biosoft/samtools-1.9/bin:$PATH&quot; &#x27;</span> &gt;&gt;~/.bashrc<br><span class="hljs-built_in">source</span> ~/.bashrc<br></code></pre></td></tr></table></figure>

<h3 id="使用-1"><a href="#使用-1" class="headerlink" title="使用"></a>使用</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-variable">$samtools</span><br><br>Program: samtools (Tools <span class="hljs-keyword">for</span> alignments <span class="hljs-keyword">in</span> the SAM format)<br>Version: 1.9 (using htslib 1.9)<br><br>Usage:   samtools &lt;<span class="hljs-built_in">command</span>&gt; [options]<br><br>Commands:<br>  -- Indexing<br>     dict           create a sequence dictionary file<br>     faidx          index/extract FASTA<br>     fqidx          index/extract FASTQ<br>     index          index alignment<br><br>  -- Editing<br>     calmd          recalculate MD/NM tags and <span class="hljs-string">&#x27;=&#x27;</span> bases<br>     fixmate        fix mate information<br>     reheader       replace BAM header<br>     targetcut      cut fosmid regions (<span class="hljs-keyword">for</span> fosmid pool only)<br>     addreplacerg   adds or replaces RG tags<br>     markdup        mark duplicates<br><br>  -- File operations<br>     collate        shuffle and group alignments by name<br>     cat            concatenate BAMs<br>     merge          merge sorted alignments<br>     mpileup        multi-way pileup<br>     sort           sort alignment file<br>     split          splits a file by <span class="hljs-built_in">read</span> group<br>     quickcheck     quickly check <span class="hljs-keyword">if</span> SAM/BAM/CRAM file appears intact<br>     fastq          converts a BAM to a FASTQ<br>     fasta          converts a BAM to a FASTA<br><br>  -- Statistics<br>     bedcov         <span class="hljs-built_in">read</span> depth per BED region<br>     depth          compute the depth<br>     flagstat       simple stats<br>     idxstats       BAM index stats<br>     phase          phase heterozygotes<br>     stats          generate stats (former bamcheck)<br><br>  -- Viewing<br>     flags          explain BAM flags<br>     tview          text alignment viewer<br>     view           SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion<br>     depad          convert padded BAM to unpadded BAM<br></code></pre></td></tr></table></figure>



<ol>
<li>view</li>
</ol>
<figure class="highlight gcode"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs gcode">view命令的主要功能是：将sam文件与bam文件互换；然后对bam文件进行各种操作，比如数据的排序<span class="hljs-comment">(sort)</span>和提取<span class="hljs-comment">(这些操作是对bam文件进行的，因而当输入为sam文件的时候，不能进行该操作)</span>；最后将排序或提取得到的数据输出为bam或sam（默认的）格式。<br>bam文件优点：bam文件为二进制文件，占用的磁盘空间比sam文本文件小；利用bam二进制文件的运算速度快。<br>view命令中，对sam文件头部（序列ID）的输入<span class="hljs-comment">(-t或-T）和输出(-h)</span>是单独的一些参数来控制的。<br></code></pre></td></tr></table></figure>

<blockquote>
<p>Usage: samtools view [options] | [region1 […]]<br> 下面的view命令的部分参数<br> 默认情况下不加 region，则是输出所有的 region.options:</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs bash">-b output BAM  <br><span class="hljs-comment"># 该参数设置输出 BAM 格式，默认下输出是 SAM 格式文件</span><br>-h <span class="hljs-built_in">print</span> header <span class="hljs-keyword">for</span> the SAM output <br><span class="hljs-comment"># 默认下输出的 sam 格式文件不带 header，该参数设定输出sam文件时带 header 信息 </span><br>-H <span class="hljs-built_in">print</span> SAM header only (no alignments)  <br><span class="hljs-comment"># 仅仅输出文件的头文件</span><br>-S input is SAM  <br><span class="hljs-comment"># 默认下输入是 BAM 文件，若是输入是 SAM 文件，则最好加该参数，否则有时候会报错。 </span><br>-u uncompressed BAM output (force -b)  <br><span class="hljs-comment"># 该参数的使用需要有-b参数，能节约时间，但是需要更多磁盘空间。 </span><br>-c <span class="hljs-built_in">print</span> only the count of matching records<br><span class="hljs-comment"># 仅输出匹配的统计记录 </span><br>-L FILE  only include reads overlapping this BED FILE [null] <br><span class="hljs-comment">#  仅包括和bed文件存在overlap的reads</span><br>-o FILE  output file name [stdout] <br><span class="hljs-comment"># 输出文件的名称</span><br>-F INT  only include reads with none of the FLAGS <span class="hljs-keyword">in</span> INT present [0] <br><span class="hljs-comment"># 过滤flag，仅输出指定FLAG值的序列</span><br>-q INT   only include reads with mapping quality &gt;= INT [0]    <br><span class="hljs-comment"># 比对的最低质量值，一般认为20就为unique比对了，可以结合上述-bF参数使用使用提取特定的比对结果</span><br>-@ Number of additional threads to use [0]<br><span class="hljs-comment"># 指使用的线程数</span><br></code></pre></td></tr></table></figure>
<p>所有例子在<code>example</code>文件下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 将sam文件转换成bam文件</span><br>samtools view -bS abc.sam &gt; abc.bam<br><br><span class="hljs-comment"># BAM转换为SAM</span><br>samtools view -h -o out.sam out.bam<br><br><span class="hljs-comment"># 提取比对到参考序列上的比对结果</span><br>samtools view -bF 4 abc.bam &gt; abc.F.bam<br><br><span class="hljs-comment"># 提取paired reads中两条reads都比对到参考序列上的比对结果，只需要把两个4+8的值12作为过滤参数即可</span><br>samtools view -bF 12 abc.bam &gt; abc.F12.bam<br><br><span class="hljs-comment"># 提取没有比对到参考序列上的比对结果</span><br>samtools view -bf 4 abc.bam &gt; abc.f.bam<br><br><span class="hljs-comment"># 提取bam文件中比对到caffold1上的比对结果，并保存到sam文件格式</span><br>samtools view abc.bam scaffold1 &gt; scaffold1.sam<br><br><span class="hljs-comment"># 提取scaffold1上能比对到30k到100k区域的比对结果</span><br>samtools view abc.bam scaffold1:30000-100000 <span class="hljs-variable">$gt</span>; scaffold1_30k-100k.sam<br><br><span class="hljs-comment"># 根据fasta文件，将 header 加入到 sam 或 bam 文件中</span><br>samtools view -T genome.fasta -h scaffold1.sam &gt; scaffold1.h.sam<br></code></pre></td></tr></table></figure>

<ol start="2">
<li>sort<blockquote>
<p>sort对bam文件进行排序。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><code class="hljs bash">Usage: samtools sort [option] &lt;in.bam&gt; -o &lt;out.prefix&gt;  <br>-n Sort by <span class="hljs-built_in">read</span> name<br><span class="hljs-comment">#设定排序方式按short reads的ID排序。默认下是按序列在fasta文件中的顺序（即header）和序列从左往右的位点排序。</span><br><br>-m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]<br><span class="hljs-comment"># 设置每个线程的最大内存，单位可以是K/M/G，默认是 768M。对于处理大数据时，如果内存够用，则设置大点的值，以节约时间。</span><br><br>-t TAG     Sort by value of TAG. Uses position as secondary index (or <span class="hljs-built_in">read</span> name <span class="hljs-keyword">if</span> -n is <span class="hljs-built_in">set</span>)<br><span class="hljs-comment"># 按照TAG值排序</span><br><br>-o FILE    Write final output to FILE rather than standard output <br><span class="hljs-comment"># 输出到文件中，加文件名</span><br></code></pre></td></tr></table></figure></li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment">#  tmp.bam 按照序列名称排序，并将结果输出到tmp.sort.bam  </span><br>samtools sort -n tmp.bam  -o tmp.sort.bam    <br>samtools view tmp.sort.bam <br></code></pre></td></tr></table></figure>

<ol start="3">
<li>merge &amp; cat</li>
</ol>
<blockquote>
<p>merge将多个已经sort了的bam文件融合成一个bam文件。融合后的文件不需要则是已经sort过了的。而cat命令不需要将bam文件进行sort。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs bash">Usage: samtools merge [-nurlf] [-h inh.sam] [-b &lt;bamlist.fofn&gt;] &lt;out.bam&gt; &lt;in1.bam&gt; [&lt;in2.bam&gt; ... &lt;inN.bam&gt;]<br><br>Options: <br>  -n         Input files are sorted by <span class="hljs-built_in">read</span> name<br><span class="hljs-comment"># 输入文件是经过sort -n的</span><br>  -t TAG     Input files are sorted by TAG value<br><span class="hljs-comment"># 输入文件是经过sort -t的</span><br>  -r         Attach RG tag (inferred from file names)<br><span class="hljs-comment"># 添加上RG标签</span><br>  -u         Uncompressed BAM output<br><span class="hljs-comment"># 输出未压缩的bam</span><br>  -f         Overwrite the output BAM <span class="hljs-keyword">if</span> exist<br><span class="hljs-comment"># 覆盖已经存在的bam</span><br>  -1         Compress level 1<br><span class="hljs-comment"># 1倍压缩</span><br>  -l INT     Compression level, from 0 to 9 [-1]<br><span class="hljs-comment"># 指定压缩倍数</span><br>  -R STR     Merge file <span class="hljs-keyword">in</span> the specified region STR [all]<br>  -h FILE    Copy the header <span class="hljs-keyword">in</span> FILE to &lt;out.bam&gt; [in1.bam]<br></code></pre></td></tr></table></figure>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-variable">$samtools</span> cat<br>Usage: samtools cat [options] &lt;in1.bam&gt;  [... &lt;inN.bam&gt;]<br>       samtools cat [options] &lt;in1.cram&gt; [... &lt;inN.cram&gt;]<br><br>Options: -b FILE  list of input BAM/CRAM file names, one per line<br>         -h FILE  copy the header from FILE [default is 1st input file]<br>         -o FILE  output BAM/CRAM<br></code></pre></td></tr></table></figure>

<ol start="4">
<li>index</li>
</ol>
<blockquote>
<p>对排序后的序列建立索引，并输出为bai文件，用于快速随机处理。在很多情况下，特别是需要显示比对序列的时候，bai文件是必不可少的，例如之后的tview命令。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools index &lt;in.bam&gt; [out.index]<br><br>samtools index abc.sort.bam<br></code></pre></td></tr></table></figure>

<ol start="5">
<li>faidx</li>
</ol>
<blockquote>
<p>对<code>fasta</code>文件建立索引,生成的索引文件以<code>.fai</code>后缀结尾。该命令也能依据索引文件快速提取<code>fasta</code>文件中的某一条（子）序列。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools faidx &lt;file.fa|file.fa.gz&gt; [&lt;reg&gt; [...]]samtools faidx &lt;file.fa|file.fa.gz&gt; [&lt;reg&gt; [...]]<br></code></pre></td></tr></table></figure>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools faidx genome.fasta<br><span class="hljs-comment"># 生成了索引文件genome.fasta.fai,是一个文本文件，分成了5列。第一列是子序列的名称；第二列是子序列的长度；个人认为“第三列是序列所在的位置”，因为该数字从上往下逐渐变大，最后的数字是genome.fasta文件的大小；第4和5列不知是啥意思。于是通过此文件，可以定位子序列在fasta文件在磁盘上的存放位置，直接快速调出子序列。</span><br><br><span class="hljs-comment"># 由于有索引文件，可以使用以下命令很快从基因组中提取到fasta格式的子序列</span><br>samtools faidx genome.fasta scffold_10 &gt; scaffold_10.fasta<br></code></pre></td></tr></table></figure>

<ol start="6">
<li>tview</li>
</ol>
<blockquote>
<p>能直观的显示出reads比对基因组的情况，和基因组浏览器有点类似。可视化一般用IGV比较好，不建议用tview</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools tview &lt;aln.bam&gt; [ref.fasta]<br></code></pre></td></tr></table></figure>

<h3 id="其他命令"><a href="#其他命令" class="headerlink" title="其他命令"></a>其他命令</h3><p>请看帮助文档<a target="_blank" rel="noopener" href="http://www.htslib.org/doc/samtools.html"><em>samtools</em></a></p>

            </div>
            <hr>
            <div>
              <div class="post-metas mb-3">
                
                  <div class="post-meta mr-3">
                    <i class="iconfont icon-category"></i>
                    
                      <a class="hover-with-bg" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a>
                    
                      <a class="hover-with-bg" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/%E6%AF%94%E5%AF%B9/">比对</a>
                    
                  </div>
                
                
                  <div class="post-meta">
                    <i class="iconfont icon-tags"></i>
                    
                      <a class="hover-with-bg" href="/tags/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a>
                    
                      <a class="hover-with-bg" href="/tags/%E6%AF%94%E5%AF%B9/">比对</a>
                    
                  </div>
                
              </div>
              
                <p class="note note-warning">
                  
                    本博客所有文章除特别声明外，均采用 <a target="_blank" href="https://creativecommons.org/licenses/by-sa/4.0/deed.zh" rel="nofollow noopener noopener">CC BY-SA 4.0 协议</a> ，转载请注明出处！
                  
                </p>
              
              
                <div class="post-prevnext">
                  <article class="post-prev col-6">
                    
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2021/07/05/PBS-use/">
                        <span class="hidden-mobile">Linux常见操作和PBS使用</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>
    
      <div class="d-none d-lg-block col-lg-2 toc-container" id="toc-ctn">
        <div id="toc">
  <p class="toc-header"><i class="iconfont icon-list"></i>&nbsp;目录</p>
  <div class="toc-body" id="toc-body"></div>
</div>

      </div>
    
  </div>
</div>

<!-- Custom -->


    

    
      <a id="scroll-top-button" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v"
                 for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>
    

    
  </main>

  <footer class="text-center mt-5 py-3">
  <div class="footer-content">
     <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
  </div>
  
  <div class="statistics">
    
    

    
      
        <!-- 不蒜子统计PV -->
        <span id="busuanzi_container_site_pv" style="display: none">
            总访问量 
            <span id="busuanzi_value_site_pv"></span>
             次
          </span>
      
      
        <!-- 不蒜子统计UV -->
        <span id="busuanzi_container_site_uv" style="display: none">
            总访客数 
            <span id="busuanzi_value_site_uv"></span>
             人
          </span>
      
    
  </div>


  

  
</footer>


  <!-- SCRIPTS -->
  
  <script  src="https://cdn.jsdelivr.net/npm/nprogress@0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/nprogress@0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" ></script>
<script  src="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>

<!-- Plugins -->


  
    <script  src="/js/img-lazyload.js" ></script>
  



  



  <script  src="https://cdn.jsdelivr.net/npm/tocbot@4.12.0/dist/tocbot.min.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.5.7/dist/jquery.fancybox.min.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/anchor-js@4.3.0/anchor.min.js" ></script>



  <script defer src="https://cdn.jsdelivr.net/npm/clipboard@2.0.6/dist/clipboard.min.js" ></script>



  <script  src="/js/local-search.js" ></script>



  <script defer src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js" ></script>




  <script  src="https://cdn.jsdelivr.net/npm/typed.js@2.0.11/lib/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var title = document.getElementById('subtitle').title;
      
      typing(title)
      
    })(window, document);
  </script>















<!-- 主题的启动项 保持在最底部 -->
<script  src="/js/boot.js" ></script>


<script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"log":false,"pluginJsPath":"lib/","pluginModelPath":"assets/","pluginRootPath":"live2dw/","tagMode":false});</script></body>
</html>
